# 📊 DecodeLabs Data Analytics Internship — Project 1
## Data Cleaning and Preparation

**Intern:** Yasir Khan  
**Date:** June 2026  
**Organization:** DecodeLabs  
**Tools:** Python, Pandas, NumPy, Jupyter Notebook

---

### 📋 Project Objective
Perform comprehensive data cleaning and preparation on a real-world e-commerce dataset to ensure data quality, consistency, and readiness for analysis.

### 📁 Dataset
- **File:** `Dataset for Data Analytics.xlsx`
- **Records:** 1,200 orders
- **Features:** 14 columns including Order ID, Date, Customer, Product, Pricing, Shipping, and Payment information


## Step 1: Import Required Libraries

In [ ]:
# ============================================================
# Step 1: Import all required libraries
# ============================================================
import pandas as pd    # Data manipulation and analysis
import numpy as np     # Numerical computing
import warnings
warnings.filterwarnings('ignore')  # Suppress warnings for cleaner output

print("✅ Libraries imported successfully")
print(f"Pandas version: {pd.__version__}")
print(f"NumPy version: {np.__version__}")

✅ Libraries imported successfully
Pandas version: 2.3.3
NumPy version: 2.3.4


## Step 2: Load the Dataset

Load the e-commerce orders dataset from the Excel file using Pandas.

In [ ]:
# ============================================================
# Step 2: Load the dataset from Excel file
# ============================================================
df = pd.read_excel("Dataset for Data Analytics.xlsx")

print(f"✅ Dataset loaded successfully!")
print(f"📏 Shape: {df.shape[0]} rows × {df.shape[1]} columns")

✅ Dataset loaded successfully!
📏 Shape: 1200 rows × 14 columns


## Step 3: Initial Data Exploration

Display the first 5 rows and dataset information to understand the structure, data types, and overall composition.

In [ ]:
# ============================================================
# Step 3a: Display first 5 rows of the dataset
# ============================================================
print("═" * 80)
print("FIRST 5 ROWS OF THE DATASET")
print("═" * 80)
df.head()

════════════════════════════════════════════════════════════════════════════════
FIRST 5 ROWS OF THE DATASET
════════════════════════════════════════════════════════════════════════════════


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


In [ ]:
# ============================================================
# Step 3b: Display dataset information (dtypes, non-null counts)
# ============================================================
print("═" * 80)
print("DATASET INFORMATION")
print("═" * 80)
df.info()
print("\n")
print("═" * 80)
print("STATISTICAL SUMMARY")
print("═" * 80)
df.describe(include="all")

════════════════════════════════════════════════════════════════════════════════
DATASET INFORMATION
════════════════════════════════════════════════════════════════════════════════
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   object        
 1   Date             1200 non-null   datetime64[ns]
 2   CustomerID       1200 non-null   object        
 3   Product          1200 non-null   object        
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   object        
 7   PaymentMethod    1200 non-null   object        
 8   OrderStatus      1200 non-null   object        
 9   TrackingNumber   1200 non-null   object        
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
count,1200,1200,1200,1200,1200.000000,1200.000000,1200,1200,1200,1200,1200.000000,891,1200,1200.000000
unique,1200,NaN,1189,7,NaN,NaN,655,5,5,1200,NaN,3,5,NaN
top,ORD200000,NaN,C38840,Printer,NaN,NaN,533 Main St,Online,Cancelled,TRK37947903,NaN,FREESHIP,Instagram,NaN
freq,1,NaN,2,181,NaN,NaN,8,258,250,1,NaN,313,259,NaN
mean,NaN,2024-03-22 16:58:48,NaN,NaN,2.945833,356.412750,NaN,NaN,NaN,NaN,5.485000,NaN,NaN,1053.968300
min,NaN,2023-01-01 00:00:00,NaN,NaN,1.000000,11.390000,NaN,NaN,NaN,NaN,1.000000,NaN,NaN,11.390000
25%,NaN,2023-08-03 18:00:00,NaN,NaN,2.000000,186.062500,NaN,NaN,NaN,NaN,4.000000,NaN,NaN,410.520000
50%,NaN,2024-03-23 00:00:00,NaN,NaN,3.000000,364.210000,NaN,NaN,NaN,NaN,5.000000,NaN,NaN,823.615000
75%,NaN,2024-11-08 12:00:00,NaN,NaN,4.000000,521.570000,NaN,NaN,NaN,NaN,7.000000,NaN,NaN,1578.475000
max,NaN,2025-06-30 00:00:00,NaN,NaN,5.000000,699.930000,NaN,NaN,NaN,NaN,10.000000,NaN,NaN,3456.400000


In [ ]:
# ============================================================
# Step 3c: Display column names and data types
# ============================================================
print("═" * 80)
print("COLUMN DETAILS")
print("═" * 80)
for i, (col, dtype) in enumerate(df.dtypes.items(), 1):
    null_count = df[col].isnull().sum()
    unique_count = df[col].nunique()
    print(f"  {i:2d}. {col:<20s} | Type: {str(dtype):<15s} | Nulls: {null_count:<5d} | Unique: {unique_count}")
print(f"\n📊 Total Columns: {len(df.columns)}")

════════════════════════════════════════════════════════════════════════════════
COLUMN DETAILS
════════════════════════════════════════════════════════════════════════════════
   1. OrderID              | Type: object          | Nulls: 0     | Unique: 1200
   2. Date                 | Type: datetime64[ns]  | Nulls: 0     | Unique: 671
   3. CustomerID           | Type: object          | Nulls: 0     | Unique: 1189
   4. Product              | Type: object          | Nulls: 0     | Unique: 7
   5. Quantity             | Type: int64           | Nulls: 0     | Unique: 5


   6. UnitPrice            | Type: float64         | Nulls: 0     | Unique: 1193
   7. ShippingAddress      | Type: object          | Nulls: 0     | Unique: 655
   8. PaymentMethod        | Type: object          | Nulls: 0     | Unique: 5
   9. OrderStatus          | Type: object          | Nulls: 0     | Unique: 5
  10. TrackingNumber       | Type: object          | Nulls: 0     | Unique: 1200
  11. ItemsInCart          | Type: int64           | Nulls: 0     | Unique: 10
  12. CouponCode           | Type: object          | Nulls: 309   | Unique: 3
  13. ReferralSource       | Type: object          | Nulls: 0     | Unique: 5
  14. TotalPrice           | Type: float64         | Nulls: 0     | Unique: 1195

📊 Total Columns: 14


## Step 4: Identify Missing / Null Values

Use `isnull().sum()` to detect missing values in each column. This is critical to determine which columns need imputation.

In [ ]:
# ============================================================
# Step 4: Identify all missing/null values using isnull().sum()
# ============================================================
missing_values = df.isnull().sum()
missing_percent = (df.isnull().sum() / len(df) * 100).round(2)

# Create a summary DataFrame for missing values
missing_df = pd.DataFrame({
    "Column": missing_values.index,
    "Missing Count": missing_values.values,
    "Missing %": missing_percent.values
})

# Sort by missing count descending
missing_df = missing_df.sort_values("Missing Count", ascending=False).reset_index(drop=True)

print("═" * 60)
print("MISSING VALUES ANALYSIS")
print("═" * 60)
print(missing_df.to_string(index=False))
print(f"\n🔍 Total missing values: {df.isnull().sum().sum()}")
print(f"📌 Columns with missing data: {(missing_values > 0).sum()}")

════════════════════════════════════════════════════════════
MISSING VALUES ANALYSIS
════════════════════════════════════════════════════════════
         Column  Missing Count  Missing %
     CouponCode            309      25.75
        OrderID              0       0.00
           Date              0       0.00
     CustomerID              0       0.00
        Product              0       0.00
       Quantity              0       0.00
      UnitPrice              0       0.00
ShippingAddress              0       0.00
  PaymentMethod              0       0.00
    OrderStatus              0       0.00
 TrackingNumber              0       0.00
    ItemsInCart              0       0.00
 ReferralSource              0       0.00
     TotalPrice              0       0.00

🔍 Total missing values: 309
📌 Columns with missing data: 1


## Step 5: Handle Missing Values

### Strategy:
- **CouponCode** (Categorical): Fill with **Mode** (most frequent value)
  - Mode is the appropriate technique for categorical data as mean/median are not applicable to non-numeric values.

> **Note:** All numerical columns (Quantity, UnitPrice, ItemsInCart, TotalPrice) have zero missing values, so mean/median imputation is not required. However, we demonstrate the approach below for completeness.

In [ ]:
# ============================================================
# Step 5: Handle missing values using appropriate techniques
# ============================================================

# --- Technique Demonstration for Numerical Columns ---
# Although no numerical columns have missing values in this dataset,
# we demonstrate the techniques that WOULD be used:

numerical_cols = df.select_dtypes(include=["int64", "float64"]).columns
print("═" * 60)
print("NUMERICAL COLUMNS - MISSING VALUE CHECK")
print("═" * 60)
for col in numerical_cols:
    skewness = df[col].skew()
    technique = "Median (skewed)" if abs(skewness) > 1 else "Mean (normal)"
    print(f"  {col:<15s} | Missing: {df[col].isnull().sum():<5d} | Skewness: {skewness:.2f} | Recommended: {technique}")

print("\n✅ No numerical columns have missing values — no imputation needed.")
print("   If they did: Mean for normally distributed, Median for skewed data.\n")

# --- Handle Categorical Missing Values ---
print("═" * 60)
print("CATEGORICAL COLUMNS - MISSING VALUE HANDLING")
print("═" * 60)

# CouponCode: Fill with Mode (most frequent value)
coupon_mode = df["CouponCode"].mode()[0]
print(f"\n  CouponCode:")
print(f"    Missing values: {df['CouponCode'].isnull().sum()}")
print(f"    Mode (most frequent): '{coupon_mode}'")

df["CouponCode"] = df["CouponCode"].fillna(coupon_mode)

print(f"    After fillna: {df['CouponCode'].isnull().sum()} missing values")
print(f"\n✅ Missing values handled successfully using Mode for categorical data!")

════════════════════════════════════════════════════════════
NUMERICAL COLUMNS - MISSING VALUE CHECK
════════════════════════════════════════════════════════════
  Quantity        | Missing: 0     | Skewness: 0.03 | Recommended: Mean (normal)
  UnitPrice       | Missing: 0     | Skewness: -0.03 | Recommended: Mean (normal)
  ItemsInCart     | Missing: 0     | Skewness: 0.00 | Recommended: Mean (normal)
  TotalPrice      | Missing: 0     | Skewness: 0.89 | Recommended: Mean (normal)

✅ No numerical columns have missing values — no imputation needed.
   If they did: Mean for normally distributed, Median for skewed data.

════════════════════════════════════════════════════════════
CATEGORICAL COLUMNS - MISSING VALUE HANDLING
════════════════════════════════════════════════════════════

  CouponCode:
    Missing values: 309
    Mode (most frequent): 'FREESHIP'
    After fillna: 0 missing values

✅ Missing values handled successfully using Mode for categorical data!


## Step 6: Identify Duplicate Rows

Use `duplicated().sum()` to check for any duplicate records in the dataset.

In [ ]:
# ============================================================
# Step 6: Identify duplicate rows using duplicated().sum()
# ============================================================
duplicate_count = df.duplicated().sum()

print("═" * 60)
print("DUPLICATE ROWS ANALYSIS")
print("═" * 60)
print(f"  Total duplicate rows found: {duplicate_count}")
print(f"  Total rows in dataset: {len(df)}")
print(f"  Percentage of duplicates: {(duplicate_count/len(df)*100):.2f}%")

if duplicate_count == 0:
    print("\n✅ No duplicate rows detected in the dataset.")
else:
    print(f"\n⚠️  {duplicate_count} duplicate rows detected — will be removed in next step.")

════════════════════════════════════════════════════════════
DUPLICATE ROWS ANALYSIS
════════════════════════════════════════════════════════════
  Total duplicate rows found: 0
  Total rows in dataset: 1200
  Percentage of duplicates: 0.00%

✅ No duplicate rows detected in the dataset.


## Step 7: Remove Duplicate Rows

Use `drop_duplicates()` to remove any duplicate records and ensure data integrity.

In [ ]:
# ============================================================
# Step 7: Remove duplicate rows using drop_duplicates()
# ============================================================
rows_before = len(df)
df = df.drop_duplicates()
rows_after = len(df)
rows_removed = rows_before - rows_after

print("═" * 60)
print("DUPLICATE REMOVAL RESULTS")
print("═" * 60)
print(f"  Rows before: {rows_before}")
print(f"  Rows after:  {rows_after}")
print(f"  Rows removed: {rows_removed}")
print(f"\n✅ drop_duplicates() executed — {rows_removed} duplicate(s) removed.")
print(f"   Dataset now has {rows_after} clean rows.")

════════════════════════════════════════════════════════════
DUPLICATE REMOVAL RESULTS
════════════════════════════════════════════════════════════
  Rows before: 1200
  Rows after:  1200
  Rows removed: 0

✅ drop_duplicates() executed — 0 duplicate(s) removed.
   Dataset now has 1200 clean rows.


## Step 8: Correct Data Formats

Ensure all data is properly formatted:
- ✅ Convert dates to **YYYY-MM-DD** format
- ✅ Remove **extra spaces** from text columns
- ✅ Standardize **text capitalization** (Title Case)
- ✅ Ensure **numerical columns** have correct data types

In [ ]:
# ============================================================
# Step 8a: Convert Date column to YYYY-MM-DD string format
# ============================================================
print("═" * 60)
print("DATE FORMAT CORRECTION")
print("═" * 60)
print(f"  Before: {df['Date'].dtype}")
print(f"  Sample (before): {df['Date'].iloc[0]}")

# Ensure Date is datetime then format to YYYY-MM-DD
df["Date"] = pd.to_datetime(df["Date"])
df["Date"] = df["Date"].dt.strftime("%Y-%m-%d")

print(f"  After:  {df['Date'].dtype}")
print(f"  Sample (after):  {df['Date'].iloc[0]}")
print(f"\n✅ Dates converted to YYYY-MM-DD format.")

════════════════════════════════════════════════════════════
DATE FORMAT CORRECTION
════════════════════════════════════════════════════════════
  Before: datetime64[ns]
  Sample (before): 2023-01-04 00:00:00
  After:  object
  Sample (after):  2023-01-04

✅ Dates converted to YYYY-MM-DD format.


In [ ]:
# ============================================================
# Step 8b: Remove extra spaces from text columns
# ============================================================
print("═" * 60)
print("TEXT CLEANING - REMOVE EXTRA SPACES")
print("═" * 60)

text_columns = df.select_dtypes(include="object").columns
print(f"  Text columns identified: {list(text_columns)}\n")

for col in text_columns:
    # Count values with leading/trailing whitespace before cleaning
    space_count = df[col].apply(lambda x: x != x.strip() if isinstance(x, str) else False).sum()
    # Strip leading and trailing spaces
    df[col] = df[col].str.strip()
    # Also replace multiple internal spaces with single space
    df[col] = df[col].str.replace(r"\s+", " ", regex=True)
    print(f"  {col:<20s} | Spaces fixed: {space_count}")

print(f"\n✅ Extra spaces removed from all {len(text_columns)} text columns.")

════════════════════════════════════════════════════════════
TEXT CLEANING - REMOVE EXTRA SPACES
════════════════════════════════════════════════════════════
  Text columns identified: ['OrderID', 'Date', 'CustomerID', 'Product', 'ShippingAddress', 'PaymentMethod', 'OrderStatus', 'TrackingNumber', 'CouponCode', 'ReferralSource']

  OrderID              | Spaces fixed: 0
  Date                 | Spaces fixed: 0
  CustomerID           | Spaces fixed: 0
  Product              | Spaces fixed: 0
  ShippingAddress      | Spaces fixed: 0
  PaymentMethod        | Spaces fixed: 0
  OrderStatus          | Spaces fixed: 0
  TrackingNumber       | Spaces fixed: 0
  CouponCode           | Spaces fixed: 0
  ReferralSource       | Spaces fixed: 0

✅ Extra spaces removed from all 10 text columns.


In [ ]:
# ============================================================
# Step 8c: Standardize text capitalization (Title Case)
# ============================================================
print("═" * 60)
print("TEXT CAPITALIZATION STANDARDIZATION")
print("═" * 60)

# Apply Title Case to selected categorical columns
capitalize_cols = ["Product", "PaymentMethod", "OrderStatus", "ReferralSource", "CouponCode"]

for col in capitalize_cols:
    df[col] = df[col].str.title()
    sample_vals = df[col].unique()[:3]
    print(f"  {col:<20s} | Sample values: {list(sample_vals)}")

print(f"\n✅ Text capitalization standardized to Title Case.")

════════════════════════════════════════════════════════════
TEXT CAPITALIZATION STANDARDIZATION
════════════════════════════════════════════════════════════
  Product              | Sample values: ['Monitor', 'Phone', 'Tablet']
  PaymentMethod        | Sample values: ['Debit Card', 'Online', 'Credit Card']
  OrderStatus          | Sample values: ['Shipped', 'Cancelled', 'Returned']
  ReferralSource       | Sample values: ['Instagram', 'Referral', 'Email']
  CouponCode           | Sample values: ['Save10', 'Freeship', 'Winter15']

✅ Text capitalization standardized to Title Case.


In [ ]:
# ============================================================
# Step 8d: Ensure numerical columns have correct data types
# ============================================================
print("═" * 60)
print("NUMERICAL DATA TYPE VERIFICATION")
print("═" * 60)

# Verify and fix numerical columns
numerical_checks = {
    "Quantity": "int64",
    "UnitPrice": "float64",
    "ItemsInCart": "int64",
    "TotalPrice": "float64"
}

for col, expected_type in numerical_checks.items():
    current_type = str(df[col].dtype)
    if current_type != expected_type:
        df[col] = df[col].astype(expected_type)
        status = "🔧 Converted"
    else:
        status = "✅ Correct"
    print(f"  {col:<15s} | Expected: {expected_type:<10s} | Current: {current_type:<10s} | {status}")

print(f"\n✅ All numerical columns verified with correct data types.")

════════════════════════════════════════════════════════════
NUMERICAL DATA TYPE VERIFICATION
════════════════════════════════════════════════════════════
  Quantity        | Expected: int64      | Current: int64      | ✅ Correct
  UnitPrice       | Expected: float64    | Current: float64    | ✅ Correct
  ItemsInCart     | Expected: int64      | Current: int64      | ✅ Correct
  TotalPrice      | Expected: float64    | Current: float64    | ✅ Correct

✅ All numerical columns verified with correct data types.


## Step 9: Final Verification

Verify that all cleaning operations were successful:
- ✅ Zero missing values
- ✅ Zero duplicate rows
- ✅ Dates correctly formatted (YYYY-MM-DD)
- ✅ All data types correct

In [ ]:
# ============================================================
# Step 9: Comprehensive Final Verification
# ============================================================
print("╔" + "═" * 58 + "╗")
print("║" + " FINAL DATA QUALITY VERIFICATION REPORT ".center(58) + "║")
print("╠" + "═" * 58 + "╣")

# Check 1: Missing Values
missing_total = df.isnull().sum().sum()
status_1 = "✅ PASS" if missing_total == 0 else "❌ FAIL"
print(f"║  1. Missing Values:    {missing_total:<10d}          {status_1:<10s}  ║")

# Check 2: Duplicate Rows
dup_total = df.duplicated().sum()
status_2 = "✅ PASS" if dup_total == 0 else "❌ FAIL"
print(f"║  2. Duplicate Rows:    {dup_total:<10d}          {status_2:<10s}  ║")

# Check 3: Date Format
date_check = all(df["Date"].str.match(r"^\d{4}-\d{2}-\d{2}$"))
status_3 = "✅ PASS" if date_check else "❌ FAIL"
print(f"║  3. Date Format:       {'YYYY-MM-DD':<10s}          {status_3:<10s}  ║")

# Check 4: Data Types
type_check = (df["Quantity"].dtype == "int64" and 
              df["UnitPrice"].dtype == "float64" and
              df["TotalPrice"].dtype == "float64" and
              df["ItemsInCart"].dtype == "int64")
status_4 = "✅ PASS" if type_check else "❌ FAIL"
print(f"║  4. Data Types:        {'Correct':<10s}          {status_4:<10s}  ║")

print("╠" + "═" * 58 + "╣")

all_pass = all([missing_total == 0, dup_total == 0, date_check, type_check])
overall = "ALL CHECKS PASSED ✅" if all_pass else "SOME CHECKS FAILED ❌"
print(f"║  {'OVERALL:':<20s} {overall:<36s}  ║")
print("╚" + "═" * 58 + "╝")

print(f"\n📊 Final Dataset Shape: {df.shape[0]} rows × {df.shape[1]} columns")

╔══════════════════════════════════════════════════════════╗
║          FINAL DATA QUALITY VERIFICATION REPORT          ║
╠══════════════════════════════════════════════════════════╣
║  1. Missing Values:    0                   ✅ PASS      ║
║  2. Duplicate Rows:    0                   ✅ PASS      ║
║  3. Date Format:       YYYY-MM-DD          ✅ PASS      ║
║  4. Data Types:        Correct             ✅ PASS      ║
╠══════════════════════════════════════════════════════════╣
║  OVERALL:             ALL CHECKS PASSED ✅                   ║
╚══════════════════════════════════════════════════════════╝

📊 Final Dataset Shape: 1200 rows × 14 columns


In [ ]:
# ============================================================
# Step 9b: Display cleaned dataset - first 5 rows
# ============================================================
print("═" * 60)
print("CLEANED DATASET — FIRST 5 ROWS")
print("═" * 60)
df.head()

════════════════════════════════════════════════════════════
CLEANED DATASET — FIRST 5 ROWS
════════════════════════════════════════════════════════════


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,Save10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,Save10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,Freeship,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,Save10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,Save10,Email,2504.04


In [ ]:
# ============================================================
# Step 9c: Display cleaned dataset information
# ============================================================
print("═" * 60)
print("CLEANED DATASET — INFORMATION")
print("═" * 60)
df.info()
print("\n")
print("═" * 60)
print("CLEANED DATASET — STATISTICAL SUMMARY")
print("═" * 60)
df.describe(include="all")

════════════════════════════════════════════════════════════
CLEANED DATASET — INFORMATION
════════════════════════════════════════════════════════════
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   OrderID          1200 non-null   object 
 1   Date             1200 non-null   object 
 2   CustomerID       1200 non-null   object 
 3   Product          1200 non-null   object 
 4   Quantity         1200 non-null   int64  
 5   UnitPrice        1200 non-null   float64
 6   ShippingAddress  1200 non-null   object 
 7   PaymentMethod    1200 non-null   object 
 8   OrderStatus      1200 non-null   object 
 9   TrackingNumber   1200 non-null   object 
 10  ItemsInCart      1200 non-null   int64  
 11  CouponCode       1200 non-null   object 
 12  ReferralSource   1200 non-null   object 
 13  TotalPrice       1200 non-null   float64
dtype

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
count,1200,1200,1200,1200,1200.000000,1200.000000,1200,1200,1200,1200,1200.000000,1200,1200,1200.000000
unique,1200,671,1189,7,NaN,NaN,655,5,5,1200,NaN,3,5,NaN
top,ORD200000,2023-08-20,C38840,Printer,NaN,NaN,533 Main St,Online,Cancelled,TRK37947903,NaN,Freeship,Instagram,NaN
freq,1,8,2,181,NaN,NaN,8,258,250,1,NaN,622,259,NaN
mean,NaN,NaN,NaN,NaN,2.945833,356.412750,NaN,NaN,NaN,NaN,5.485000,NaN,NaN,1053.968300
std,NaN,NaN,NaN,NaN,1.407557,197.177146,NaN,NaN,NaN,NaN,2.281983,NaN,NaN,819.856558
min,NaN,NaN,NaN,NaN,1.000000,11.390000,NaN,NaN,NaN,NaN,1.000000,NaN,NaN,11.390000
25%,NaN,NaN,NaN,NaN,2.000000,186.062500,NaN,NaN,NaN,NaN,4.000000,NaN,NaN,410.520000
50%,NaN,NaN,NaN,NaN,3.000000,364.210000,NaN,NaN,NaN,NaN,5.000000,NaN,NaN,823.615000
75%,NaN,NaN,NaN,NaN,4.000000,521.570000,NaN,NaN,NaN,NaN,7.000000,NaN,NaN,1578.475000


## Step 10: Save Cleaned Dataset

Export the cleaned and validated dataset as a CSV file for downstream analysis.

In [ ]:
# ============================================================
# Step 10: Save the cleaned dataset as CSV
# ============================================================
output_file = "cleaned_dataset.csv"
df.to_csv(output_file, index=False)

print("═" * 60)
print("DATASET SAVED SUCCESSFULLY")
print("═" * 60)
print(f"  📁 File: {output_file}")
print(f"  📏 Rows: {df.shape[0]}")
print(f"  📊 Columns: {df.shape[1]}")
print(f"  💾 Size: {round(df.memory_usage(deep=True).sum() / 1024, 2)} KB")
print(f"\n✅ Cleaned dataset saved as '{output_file}'")

════════════════════════════════════════════════════════════
DATASET SAVED SUCCESSFULLY
════════════════════════════════════════════════════════════
  📁 File: cleaned_dataset.csv
  📏 Rows: 1200
  📊 Columns: 14
  💾 Size: 709.53 KB

✅ Cleaned dataset saved as 'cleaned_dataset.csv'


---
## 📝 Change Log

| Change ID | Description | Impact | Status |
|-----------|------------|--------|--------|
| CL-001 | Loaded dataset from Excel file | Established raw data baseline with 1,200 records | ✅ Completed |
| CL-002 | Identified 309 missing values in CouponCode | Revealed data quality gap in coupon tracking | ✅ Completed |
| CL-003 | Filled CouponCode nulls with Mode (most frequent value) | Restored 309 records to complete status; no data loss | ✅ Completed |
| CL-004 | Checked for duplicate rows | Confirmed 0 duplicates — data integrity verified | ✅ Completed |
| CL-005 | Executed drop_duplicates() | Ensured deduplication pipeline is in place for future data | ✅ Completed |
| CL-006 | Converted Date column to YYYY-MM-DD format | Standardized date format for consistent analysis | ✅ Completed |
| CL-007 | Stripped whitespace from all text columns | Eliminated potential matching/grouping errors | ✅ Completed |
| CL-008 | Standardized text to Title Case | Ensured consistent capitalization across categorical fields | ✅ Completed |
| CL-009 | Verified numerical column data types | Confirmed int64/float64 types for accurate calculations | ✅ Completed |
| CL-010 | Final verification — all checks passed | Zero nulls, zero duplicates, correct formats confirmed | ✅ Completed |
| CL-011 | Saved cleaned dataset as cleaned_dataset.csv | Deliverable ready for downstream analysis | ✅ Completed |

---
## 🎯 Conclusion

This project successfully demonstrated a professional data cleaning and preparation workflow on a real-world e-commerce dataset.

### Key Accomplishments:
- **Missing Values:** Identified 309 null values in `CouponCode` and imputed them using the **Mode** technique (appropriate for categorical data).
- **Duplicates:** Verified zero duplicate rows and applied `drop_duplicates()` as a best practice.
- **Data Formatting:** Standardized dates to `YYYY-MM-DD`, removed whitespace, applied Title Case, and verified numerical data types.
- **Validation:** All quality checks passed — the dataset is clean, consistent, and ready for analysis.

### Tools & Technologies:
- **Python** — Programming language
- **Pandas** — Data manipulation and cleaning
- **NumPy** — Numerical operations
- **Jupyter Notebook** — Interactive development environment

---
*Project completed by Yasir Khan | DecodeLabs Data Analytics Internship | June 2026*